### PR TITLE
Add ports of Must-Fix CA IRQL rules (C28157, C28158)

### PR DESCRIPTION
**Pull request by NateD-MSFT**
[C28157: The IRQL was never restored](https://learn.microsoft.com/en-us/windows-hardware/drivers/devtest/28157-function-irql-never-restored)
[C28158: No IRQL was saved](https://learn.microsoft.com/en-us/windows-hardware/drivers/devtest/28158-no-irql-was-saved)

This PR also adds a variant of C28157 in the "experimental" folder that uses "must-flow" analysis; however, this functionality is incomplete at the moment and will require an update to a new CodeQL version and libraries.

## Checklist for Pull Requests

- [x] Description is filled out.
- [x] Only one query or related query group is in this pull request.
- [x] The version number on changed queries has been increased via the `@version` comment in the file header.
- [x] All unit tests have been run: ([Test README.md](src\drivers\test\README.md)).
- [x] Commands `codeql database create` and `codeql database analyze` have completed successfully.
- [x] A .qhelp file has been added for any new queries or updated if changes have been made to an existing query.

### CODE REVIEW

File: src/drivers/general/queries/IrqlNotSaved/IrqlNotSaved.ql [hunk 1]
```diff
@@ -0,0 +1,169 @@
+// Copyright (c) Microsoft Corporation.
```
Comment (lauragonzalezzz) [inline src/drivers/general/queries/IrqlNotSaved/IrqlNotSaved.ql:1]: Since this is a warning/low severity issue with only medium precision, I assert that this isn't quite ready to introduce into our collection yet. I recommend keeping it in the experimental directory until the full version is complete and the rule can detect if there are any code paths which don't save the IRQL.

If you open bugs with the rule as is, you have high chance of devs interpreting a duplicate bug when the rule is complete because it will cover the same code paths with inverse logic: one bug for not having a code path which saves, one bug for having a code path which doesn't save.

File: src/drivers/general/queries/IrqlNotSaved/driver_snippet.c [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+//
+// driver_snippet.c
+//
+
+#define SET_DISPATCH 1
+
+// Template. Not called in this test.
+void top_level_call() {}
+
+// Auxilliary function used in passing.
```
Comment (lauragonzalezzz) [inline src/drivers/general/queries/IrqlNotSaved/driver_snippet.c:12]: Please use */ ... /* for multiline comments.
Spelling: "though"

File: src/drivers/general/queries/IrqlNotSaved/driver_snippet.c [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+
+// Failing case one.
+// Function does nothing with IRQL.
+VOID IrqlNotSaved_fail1(_IRQL_saves_ KIRQL outIrql, PKSPIN_LOCK myLock) {
+
+}
+
+// Failing case two.
+// Function has a path where the IRQL is not saved.
+// This requies must-flow analysis.
```
Comment (lauragonzalezzz) [inline src/drivers/general/queries/IrqlNotSaved/driver_snippet.c:44]: Wouldn't this test be passing for the current iteration of this rule?

Spelling: requires
Please use a spell checker. I like [this one](https://marketplace.visualstudio.com/items?itemName=EWoodruff.VisualStudioSpellCheckerVS2022andLater).

File: src/drivers/general/queries/IrqlNotUsed/IrqlNotUsed.ql [hunk 4]
```diff
@@ -0,0 +1,94 @@
+// Copyright (c) Microsoft Corporation.
```
Comment (lauragonzalezzz) [inline src/drivers/general/queries/IrqlNotUsed/IrqlNotUsed.ql:1]: Same suggestion as previous rule.

File: src/drivers/general/queries/IrqlNotUsed/driver_snippet.c [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+
+}
+
+// Failing case one.
+// Function does nothing with IRQL.
+VOID IrqlNotUsed_fail1(_IRQL_restores_ KIRQL inIrql, PKSPIN_LOCK myLock) {
+
+}
+
+// Failing case two.
```
Comment (lauragonzalezzz) [inline src/drivers/general/queries/IrqlNotUsed/driver_snippet.c:42]: Same comment as previous rule.

File: src/drivers/general/queries/experimental/IrqlNotUsedPrecise/IrqlNotUsedPrecise.ql [hunk 6]
```diff
@@ -0,0 +1,92 @@
+// Copyright (c) Microsoft Corporation.
```
Comment (lauragonzalezzz) [inline src/drivers/general/queries/experimental/IrqlNotUsedPrecise/IrqlNotUsedPrecise.ql:1]: I believe this is in line with my previous comments, but why have a completely separate rule when we have built versioning into our queries? Let's keep it simple and only have one version at a time.

File: src/drivers/libraries/Irql.qll [hunk 7]
```diff
@@ -23,6 +23,52 @@ class IrqlTypeDefinition extends SALAnnotation {
...
+  IrqlParameterAnnotation() {
+    this.getMacroName().matches(["_IRQL_restores_", "_IRQL_saves_"]) and
+    irqlAnnotationName = this.getMacroName() and
+    exists(MacroInvocation mi | mi.getParentInvocation() = this)
+  }
+
+  string getIrqlMacroName() { result = irqlAnnotationName }
+}
+
+/**
```
Comment (lauragonzalezzz) [inline src/drivers/libraries/Irql.qll:42]: IrqlRestore and Irql Save descriptions sound like the same thing. I don't understand, can you rephrase?

File: src/drivers/test/diff/IrqlNotSaved.sarif [hunk 8]
```diff
@@ -0,0 +1,21 @@
+{
```
Comment (lauragonzalezzz) [inline src/drivers/test/diff/IrqlNotSaved.sarif:1]: Can you use this as a template instead of saving the same version of this for every query?

File: src/suites/ported_driver_ca_checks.qls [hunk 9]
```diff
@@ -5,7 +5,10 @@
   from: microsoft/windows-drivers
```
Comment (lauragonzalezzz) [inline src/suites/ported_driver_ca_checks.qls:5]: What is the purpose of putting all ported queries in a suite?

**Review by lauragonzalezzz — CHANGES_REQUESTED** [review]

File: src/drivers/general/queries/IrqlNotSaved/IrqlNotSaved.ql (hunk 1 above)
Comment (NateD-MSFT) [reply src/drivers/general/queries/IrqlNotSaved/IrqlNotSaved.ql:1]: FWIW, I selected "medium" here not because of false positives, but because of false negatives.  I expect this rule to miss (and it does miss) some cases due to the lack of must-flow analysis.  But I expect the false-positive rate to be low.  Do we want to use only false positive rate to determine if something is "experimental?"

You do make a good point about potential dev confusion if we upgrade the rule and file additional bugs.

File: src/drivers/general/queries/IrqlNotSaved/driver_snippet.c (hunk 3 above)
Comment (NateD-MSFT) [reply src/drivers/general/queries/IrqlNotSaved/driver_snippet.c:44]: Yes, this case does (incorrectly) pass.  It's named after the expected result, not the actual result.  

Thanks for the spell checker link, but this code is written in VS Code - I'll look to see if there's something similar for that.

File: src/drivers/general/queries/experimental/IrqlNotUsedPrecise/IrqlNotUsedPrecise.ql (hunk 6 above)
Comment (NateD-MSFT) [reply src/drivers/general/queries/experimental/IrqlNotUsedPrecise/IrqlNotUsedPrecise.ql:1]: In this case, it's because I built this extended version before learning that the must-flow analysis isn't working right now.  Rather than deal with keeping track of this version somewhere private, I figured I may as well include it as "experimental".  I could alternatively create a branch with this version as an update to the other rule, which would sit open until we could get updated QLPacks/etc.

File: src/drivers/libraries/Irql.qll (hunk 7 above)
Comment (NateD-MSFT) [reply src/drivers/libraries/Irql.qll:42]: I'll try and rephrase a bit.  Essentially:
irql_restores: The value in this variable will be used to set the system IRQL.  (The variable is read from.)
irql_saves: The system will save the current IRQL to this variable.  (The variable is written to.)

File: src/drivers/test/diff/IrqlNotSaved.sarif (hunk 8 above)
Comment (NateD-MSFT) [reply src/drivers/test/diff/IrqlNotSaved.sarif:1]: These files are a way for us to keep track of "baseline debt".  The SARIF files checked in are the expected results for running our tests today.  If a change happens that results in a test result changing, we would save that diff to this file for that rule.  ATM they're redundant because we haven't had a regression like that yet.

File: src/suites/ported_driver_ca_checks.qls (hunk 9 above)
Comment (NateD-MSFT) [reply src/suites/ported_driver_ca_checks.qls:5]: In addition to being helpful as a quick way to check what's been ported, it'll be useful when we do wider CA vs CodeQL analyses to be able to run all our ported queries together without dealing with the noise of running unrelated queries.

**Review by lauragonzalezzz — APPROVED** [review]
